src/ws/messageRouter.ts

```diff
@@ -0,0 +1,157 @@
+/** Event subscribing / routing for recieved websocket messages */
+
+import { randomUUID } from "node:crypto";
+import { Logger } from "../logging";
+import { Message, MessageType } from "./messageTypes";
+
+/** Type describing message handler callbacks to whom messages are routed. */
+export type MessageCallback<T extends MessageType> = (message: Message<T>) => Promise<void>;
+
+/** Type keeping track of a per-message-type callbacks. */
+export type CallbackEntry<T extends MessageType> = {
+  callback: MessageCallback<T>; // Callback to be called when a message of this type is recieved.
+  once: boolean; // Indicates if the callback should be called only once, then removed. This featurette might never be used.
+  registrationToken: string; // Unique token to identify the callback for removal.
+};
+
+/** Type for a map of message type -> array of callback entries.*/
+export type CallbackMap = Map<MessageType, CallbackEntry<any>[]>;
+
+const logger = new Logger("messageRouter");
+
+export class MessageRouter {
```

Comment:
I get that this class is essentially a event emitter _pattern_ implementation, but I'd still suggest looking into how we can utilize Web's native [EventTarget](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget) or Node.js [EventEmitter](https://nodejs.org/en/learn/asynchronous-work/the-nodejs-event-emitter).

Reply:
The main reason I suggest looking into these is that both implementation have convenient interfaces for one time subscriptions:

```js
let events = new EventTarget();

events.addEventListener("something", handler, { once: true });
```

```js
let events = new EventEmitter();

events.once("something", handler);
```

Reply:
Ok, meet you in the middle and implement MessageRouter on top of a map of per-event-type EventEmitters? That'd reduce the LOC while keeping the existing external API.